plugins/ssl/certificate_file_expiry

```diff
@@ -72,23 +87,70 @@ if [ "$1" = "config" ] ; then
   echo "graph_category security"
 fi
 
+if [ "${warning:-}" = "" ] ; then
+  # as default when certificate is only valid for 5 days or less emit a warning
+  warning="5:"
+fi
+if [ "${critical:-}" = "" ] ; then
+  # as default when certificate is only valid for 1 day or less emit a critical
+  critical="1:"
+fi
+
 now=$(date +%s)
+get_validity() {
+  openssl_type=$1
+  file=$2
+  if [ "$file" != "--" ] ; then
+    validity=$(/usr/bin/openssl "$openssl_type" -text -noout -in "$file" | grep -E '(Next Update|Not After)')
+  else
+    # when file is set to -- read from stdin
+    validity=$(/usr/bin/openssl "$openssl_type" -text -noout | grep -E '(Next Update|Not After)')
+  fi
+  validity=${validity#*:}
+  validity=$(date --date="$validity" +%s)
+  validity=$((validity - now))
+  echo "$validity" | awk '{ print ($1 / 86400) }'
+}
+print_config_lines() {
+  name=$1
+  label=$2
+  echo "${name}.label ${label}"
+  print_warning "$name"
+  print_critical "$name"
+}
+get_openvpn_inline_cert() {
+  file=$1
+  type=$2
+  # print content between <type> and </type> lines (ca and cert)
+  awk 'BEGIN{content=0}/^<\/'"$type"'>$/{content=0}(content==1){ print $0 }/^<'"$type"'>$/{content=1}' < "$file"
+}
+
 for cert in ${CERTS}; do
   cert_type=${cert%:*}
   cert_pattern=${cert#*:}
   for cert_file in $cert_pattern; do
-    cert_name=$(clean_fieldname "$cert_file")
-    if [ "$1" = "config" ] ; then
-      echo "${cert_name}.label ${cert_file}"
-      print_warning "$cert_name"
-      print_critical "$cert_name"
-    elif [ "$1" = "" ] ; then
-      validity=$(/usr/bin/openssl "$cert_type" -text -noout -in "$cert_file" | grep -E '(Next Update|Not After)')
-      validity=${validity#*:}
-      validity=$(date --date="$validity" +%s)
-      validity=$((validity - now))
-      validity=$(echo "$validity" | awk '{ print ($1 / 86400) }')
-      echo "${cert_name}.value $validity"
+    if [ "$IGNORE_UNEXPANDED_PATTERNS" = "yes" ] \
+      && [ "$cert_file" = "$cert_pattern" ] \
+      && echo "$cert_file" | grep -q "*" ; then
```

Comment:
Are you sure, that you want to `grep` for `*`?
Maybe you meant the dot instead ("any character")? Or maybe simple use `[ -n "$cert_file" ]` in this case?

Reply:
yes i'm sure (as sure as a i can be, say 95%), grepping for `*` seemed the safest and most compatible way to check if the `cert_file` variable contains an unexpanded pattern e.g. `/etc/letsencrypt/live/*/cert.pem` which is the core of the `IGNORE_UNEXPANDED_PATTERNS` flag.

this enables us to use one config for multiple nodes that have not all files present at all times (e.g. in a pre prepared docker image where files are mapped into the container or not):
```
[certificate_file_expiry]
user root
env.CERTS crl:/etc/monitor_certs/*/crl.pem x509:/etc/monitor_certs/*/*.crt openvpn_inline:/etc/monitor_certs/*/client.conf x509:/etc/monitor_certs/*/cert.pem
env.IGNORE_UNEXPANDED_PATTERNS yes
```

i added the flag for the behaviour without the flag to stay the same (showing errors) in case somebody depends on those errors (although that is totally not adviced)

i added a few lines of comments to make it clear the `grep -q "*"` is not an accident

Reply:
Thanks for the explanation. The comments will be useful.
Probably it would have been more obvious to me, if you either used the `-F` flag or escaped the star with a backslash.